rust/processor/src/bq_analytics/generic_parquet_processor.rs

```diff
@@ -153,84 +161,106 @@ where
-                .as_slice()
-                .write_to_row_group(&mut row_group_writer)
-                .unwrap();
-            row_group_writer.close()?;
-            self.close_writer()?;
-
-            debug!(
-                table_name = ParquetType::TABLE_NAME,
-                start_version = start_version,
-                end_version = end_version,
-                "Max buffer size reached, uploading to GCS."
-            );
-            let bucket_root = PathBuf::from(&self.bucket_root);
-            let upload_result = upload_parquet_to_gcs(
-                gcs_client,
-                &new_file_path,
-                ParquetType::TABLE_NAME,
-                &self.bucket_name,
-                &bucket_root,
-            )
-            .await;
-            self.buffer_size_bytes = 0;
-            remove_file(&new_file_path)?;
-
-            return match upload_result {
-                Ok(_) => {
-                    let parquet_processing_result = ParquetProcessingResult {
-                        start_version,
-                        end_version,
-                        last_transaction_timestamp: Some(last_transaction_timestamp),
-                        txn_version_to_struct_count,
-                    };
-
-                    self.gap_detector_sender
-                        .send(ProcessingResult::ParquetProcessingResult(
-                            parquet_processing_result,
-                        ))
-                        .await
-                        .expect("[Parser] Failed to send versions to gap detector");
-                    Ok(())
-                },
-                Err(e) => {
-                    error!("Failed to upload file to GCS: {}", e);
-                    Err(anyhow!("Failed to upload file to GCS: {}", e))
-                },
-            };
+            if self.buffer_size_bytes >= self.max_buffer_size {
+                info!("Max buffer size reached, uploading to GCS.");
+                if let Err(e) = self.upload_buffer(gcs_client).await {
+                    error!("Failed to upload buffer: {}", e);
+                    return Err(e);
+                }
+                self.last_upload_time = Instant::now();
+            }
+
+            if self.last_upload_time.elapsed() >= self.upload_interval {
+                info!(
+                    "Time has elapsed more than {} since last upload.",
+                    self.upload_interval.as_secs()
+                );
+                if let Err(e) = self.upload_buffer(gcs_client).await {
+                    error!("Failed to upload buffer: {}", e);
+                    return Err(e);
+                }
+                self.last_upload_time = Instant::now();
+            }
         }
 
         PARQUET_HANDLER_BUFFER_SIZE
             .with_label_values(&[ParquetType::TABLE_NAME])
             .set(self.buffer.len() as i64);
         Ok(())
     }
+
+    pub async fn upload_buffer(&mut self, gcs_client: &GCSClient) -> Result<()> {
```

Comment:
nit: remove `pub`? this leaves the `handle` to be the only operation for the buffer manager.